File: DSA(Geeks for Geeks)/Basics/Stack/Lectures/05. Balanced Parantheses.js
// Balanced Parantheses....

function matching(a, b) {
    return ((a === '(' && b === ')') || 
            (a === '{' && b === '}') ||
             a === '[' && b === ']');
}

function isBalanced(str) {
    let s = [];

    for(let i = 0; i < str.length; i++) {
        let x = str[i];
        if(x === '(' || x === '[' || x === '{') {
            s.push(x);
        } else { // closing brackets
            if(s.length === 0) {
                return false;
            } else if(!matching(s[s.length - 1], x)) {
                return false;
            } else {
                s.pop();
            }
        }
    }

    return (s.length === 0);
}


console.log(isBalanced('(){()}((()))'))

File: python/DSA(Geeks for Geeks)/Basics/Stack/05. Balanced Parantheses.py
# // Balanced Parantheses....

def matching(a, b) :
    return ((a == '(' and b == ')') or
            (a == '{' and b == '}') or 
            (a == '[' and b == ']'))

def isBalanced(str):
    s = []

    for i in range(len(str)):
        x = str[i]
        if x == '(' or x == '[' or x == '{' :
            s.append(x)
        else :
            if len(s) == 0:
                return False
            elif matching(s[len(s) - 1], x) == False:
                return False
            else:
                s.pop()
    
    return len(s) == 0


print(isBalanced('(){()}]()'))